### DISCUSSION TITLE
Extracting one child feature from an array of child features

### DISCUSSION THREAD

**Discussion by gkaimakas** [pointfreeco/swift-composable-architecture#2902] [General]
I have a parent feature, let's call it `Week`. A `Week` can have some child features. let's call them `Day`. A `Week.State` also has a `selecteDay` property that I use to pluck out one day and display to the user. My current setup looks like the following:

```Swift
@Reducer
struct Week {
    @ObservableState
    struct State {
        var selectedDay: Day.State.ID
        var days: IdentifiedArrayOf<Day.State> = .init()
    }
    
    enum Action {
        case days(IdentifiedActionOf<Day>)
    }
    
    var body: some ReducerOf<Self> {
        Reduce(core)
            .forEach(\.days, action: \.days) {
                Day()
            }
    }
    
    func core(
        state: inout State,
        action: Action
    ) -> Effect<Action> {
        .none
    }
}

public extension Store where State == Week.State, Action == Week.Action {
    func day(_id: Day.State.ID) -> Store<Day.State?, Day.Action> {
        scope(
            state: { $0.days[id: id] }
            action: { childAction -> Action in
                Action.day(.element(id, childAction))
            }
        )
    }
}
```

The way that I'm doing things though never felt elegant and with the newest versions of TCA, the scope that I'm using is deprecated and ideally I'd want to move away from this ASAP.

Is there a  better way to tackle this?

I would assume that `.filter(where: (ChildFeature.State.ID) -> Bool)` and `first(where: (ChildFeature.State.ID) -> Bool)` functions would be available for stores that the state is `IdentifiedArrayOf<Something>` but that is not the case.

**Comment by tkunst891** [comment]
Just use the new CaseKeypaths and return: `self.scope(state: \.days[id: id], action: \.days[id: id])`

**Reply by gkaimakas** [reply]
That's great! Thanks!

**Reply by TC72** [reply]
I'm writing something similar, could you explain where this would be returned?
I'm still learning TCA and trying to get my head around these kinds of things.

**Reply by gkaimakas** [reply]
@TC72 An example would be that you want to show only one Day at a time but you also want to keep any changes that happen in other Days in memory (helps a lot if you don't have the resources to persist changes).

Another place that I use this extensively is with UIKit but the uses cases there are specific to my codebase.

**Reply by tkunst891** [reply]
> I'm writing something similar, could you explain where this would be returned? I'm still learning TCA and trying to get my head around these kinds of things.

This would be returned in the `func day` function of the `extension Store` in the original question. It would return a Store which contains just one <Day> of the original store: `Store<Day.State?, Day.Action>`, but the `Day` State is optional, because the requested id might not be contained in the original collection. 

Actually a better signature for the return value of the function would be `Store<Day.State, Day.Action>?`, so the function returns an optional Store of an honest Day. This way you can check the resulting store with an `if let ` and then use the unwrapped store to give it for example to a view which requires a `StoreOf<Day>`.

**Reply by TC72** [reply]
Thanks for the explanation, that's great.
